### PR TITLE
fix: steps that should be skipped if not applicable

### PR DESCRIPTION
**Pull request by sshanzel**
## Changes
- Introduce a process where we determine which steps should automatically be skipped.
- For more info: https://dailydotdev.slack.com/archives/C02E2C3C13R/p1757472030940799
- Tested locally and worked pretty well.

## Events

Did you introduce any new tracking events?

## Experiment

Did you introduce any new experiments?

## Manual Testing

> [!CAUTION]
> Please make sure existing components are not breaking/affected by this PR

### CODE REVIEW

File: packages/shared/src/features/onboarding/hooks/useFunnelNavigation.ts [hunk 1]
```diff
@@ -87,10 +91,80 @@ function updateURLWithStepId({
   router.push(`${pathname}?${params.toString()}`, { scroll: true });
 }
 
+function resolveDestination(
+  destination: string,
+  context?: {
+    chapters: Chapters;
```
Comment (nensidosari) [inline packages/shared/src/features/onboarding/hooks/useFunnelNavigation.ts]: What is a Chapter?

File: packages/shared/src/features/onboarding/hooks/useFunnelNavigation.ts [hunk 2]
```diff
@@ -195,56 +269,21 @@ export const useFunnelNavigation = ({
...
+      transition.destination,
+      steps,
+      shouldSkipRef?.current || {},
+      context,
+    );
 
     return {
       placement: transition.placement,
-      destination: nextStep.id,
+      destination: finalDestination,
```
Comment (nensidosari) [inline packages/shared/src/features/onboarding/hooks/useFunnelNavigation.ts]: This line scared me 😅

File: packages/shared/src/features/onboarding/hooks/useFunnelNavigation.ts (hunk 1 above)
Comment (sshanzel) [reply packages/shared/src/features/onboarding/hooks/useFunnelNavigation.ts]: A funnel consists of Chapters, and Chapters consist of steps. Basically, a Cluster of steps.

File: packages/shared/src/features/onboarding/hooks/useFunnelNavigation.ts (hunk 2 above)
Comment (sshanzel) [reply packages/shared/src/features/onboarding/hooks/useFunnelNavigation.ts]: Gotta be creative with naming variables every once in a while 😂

File: packages/shared/src/features/onboarding/hooks/useFunnelNavigation.ts [hunk 3]
```diff
@@ -87,6 +90,117 @@ function updateURLWithStepId({
...
+
+  const shouldSkip = shouldSkipMap[next.type];
+
+  if (!shouldSkip) {
+    return resolvedDestination;
+  }
+
+  // step should be automatically skipped - treat as "auto-completed"
+  const completeTransition = next.transitions.find(
+    (t) => t.on === FunnelStepTransitionType.Complete,
```
Comment (rebelchris) [inline packages/shared/src/features/onboarding/hooks/useFunnelNavigation.ts:161]: This is some freya internal?
What does it mean to complete a skipped step?

Comment (sshanzel) [reply packages/shared/src/features/onboarding/hooks/useFunnelNavigation.ts:161]: I've thought of making it configurable, but there are still certain things that must be checked through the apps code still.

This logic is to handle cases where a step is not applicable due to certain conditions, for example, it must only be for mobile, or vice versa. Previously, the funnel still went through every step, then the side effects realizes that it is not applicable, the step sends a complete event, which is inherently wrong as we talked about in Slack (and as a resut, the flicker).

The logic here traverses the steps and finds the **next applicable** step. If there's a step that we hit that must be skipped due to the conditions mentioned, this will correctly search for the next/right one.

File: packages/shared/src/features/onboarding/hooks/useFunnelNavigation.ts [hunk 4]
```diff
@@ -87,6 +90,115 @@ function updateURLWithStepId({
...
+  ) {
+    const chapter = funnelChapters[chapterIndex];
+    for (let stepIndex = 0; stepIndex < chapter.steps.length; stepIndex += 1) {
+      if (chapter.steps[stepIndex].id === stepId) {
+        return { chapter: chapterIndex, step: stepIndex };
+      }
+    }
+  }
+  return null;
+}
```
Comment (ilasw) [inline packages/shared/src/features/onboarding/hooks/useFunnelNavigation.ts]: We already have the `stepMap`, using it we can access any step using it as key `stepMap[stepId]`.

implementation is 

```ts
function getStepMap(funnel: FunnelJSON): StepMap {
  const stepMap: StepMap = {};

  funnel.chapters.forEach((chapter, chapterIndex) => {
    chapter.steps.forEach((step, stepIndex) => {
      stepMap[step.id] = {
        position: {
          chapter: chapterIndex,
          step: stepIndex,
        },
      };
    });
  });

  return stepMap;
}
```

Comment (sshanzel) [reply packages/shared/src/features/onboarding/hooks/useFunnelNavigation.ts]: Oh, right. Great to know we have this already. Let me update it quickly!

File: packages/shared/src/features/onboarding/hooks/useFunnelNavigation.ts (hunk 3 above)
Comment (sshanzel) [reply packages/shared/src/features/onboarding/hooks/useFunnelNavigation.ts:161]: The test cases should carefully explain the edge cases realized.

File: packages/shared/src/features/onboarding/steps/FunnelInstallPwa.tsx [hunk 5]
```diff
@@ -43,4 +25,12 @@ function FunnelInstallPwaComponent({
   );
 }
 
-export const FunnelInstallPwa = withIsActiveGuard(FunnelInstallPwaComponent);
+export const FunnelInstallPwa = withShouldSkipStepGuard(
```
Comment (rebelchris) [inline packages/shared/src/features/onboarding/steps/FunnelInstallPwa.tsx:28]: Just idea for future, wouldn't it be useful if we could add guards to Freya?
EG:
````
guards: {
    isIOS: false,
    isPWA: true
   isMobile: false
}
```

Those could then represent back to apply the right conditions per step?
It's not for this PR, but maybe raise it as future improvement, would make things a lot more dynamic I feel like.

Comment (sshanzel) [reply packages/shared/src/features/onboarding/steps/FunnelInstallPwa.tsx:28]: I agree with that. It was initially my plan to have that too, but I realized that there are two things going on. One is the deeper issue with the flow management, and one is the configuration like you mentioned. The configuration would require a lot of changes too, and hence didn't want to combine the two changes together, and there are a few more things to consider that we will have to hard-code as well. Many things in consideration, but yes, it would be amazing if we can introduce some sort of config for this.
